src/Dependencies/PooledObjects/SpannableArrayBuilder.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.PooledObjects;
+
+/// <summary>
+/// Pooled array building data structure that allows access to the underlying array as a span before being freed back to the pool.
+/// </summary>
+[DebuggerDisplay("Count = {Count,nq}")]
+internal sealed class SpannableArrayBuilder<T>
+{
+    public const int PooledArrayLengthLimitExclusive = 128;
+
+    private static readonly ObjectPool<SpannableArrayBuilder<T>> s_pool = new(() => new(capacity: 4), size: 16);
+
+    private T[] _items;
+
+    internal SpannableArrayBuilder(int capacity = 8)
+    {
+        _items = new T[capacity];
+        Count = 0;
+    }
+
+    public static SpannableArrayBuilder<T> GetInstance()
+        => s_pool.Allocate();
+
+    public void Free()
+    {
+        if (_items.Length < PooledArrayLengthLimitExclusive)
+        {
+            if (Count != 0)
+                Clear();
+
+            s_pool.Free(this);
+        }
+        else
+        {
+            s_pool.ForgetTrackedObject(this);
+        }
+    }
+
+    public void Add(T item)
+    {
+        var newCount = Count + 1;
+        EnsureCapacity(newCount);
+        _items[Count] = item;
+        Count = newCount;
+    }
+
+    public void AddRange(ReadOnlySpan<T> items)
+    {
+        foreach (var item in items)
+            Add(item);
+    }
+
+    public void EnsureCapacity(int capacity)
+    {
+        if (_items.Length < capacity)
+        {
+            var newCapacity = Math.Max(_items.Length * 2, capacity);
+            Array.Resize(ref _items, newCapacity);
+        }
+    }
+
+    public int Count { get; private set; }
+    public int Capacity => _items.Length;
+
+    public ref T this[int index] => ref _items[index];
+
+    public void Clear()
+    {
+        Array.Clear(_items, 0, Count);
+        Count = 0;
+    }
+
+    public ReadOnlySpan<T> AsSpan()
+        => new(_items, 0, Count);
```

Comment:
> => new(_items, 0, Count); [](http://example.com/codeflow?start=7&length=26)

I am not sure whether we would want to have a collection like that.
1. The span might no longer reflect the content of the builder after `EnsureCapacity` is called. 
2. Same issue with the ref returned by the indexer.
3. When `T` is a reference type, performance characteristics for indexing could be worse than for array builder due to a type check performed by the ```ref _items[index]``` operation.

We might want to look for an alternative solution to the problem that we are trying to address.

Reply:
For example, picking either ```ArrayBuilder``` or ```ImmutableArray``` for the input.

Reply:
I view the AsSpan as similar to what ImmutableCollectionsMarshal.AsArray and CollectionsMarshal.AsArray provide. Perhaps it would be better to not put the AsSpan directly on the class, but rather in a Marshal like class to indicate that extra care need be taken while the span is in use?

Reply:
Went ahead and pushed some changes over in commit 6 to use a marshal class mechanism similar to what List/ImmutableArray and our Segmented collections do. I don't think SpannableArrayBuilder remains a good name for the collection class, but I couldn't think of a more applicable name right now.

Reply:
> but also allowed us to generally solve the problem of being able to call span based APIs from callers who need ArrayBuilder like behaviors.

Why ```List<T>```, would not work? I am still not sure why do we need to invent a new private collection implementation

Reply:
A couple thoughts that make List less appealing:
1) I don't think the ability to get the underlying array from a List is exposed on netfx
2) List doesn't directly support pooling
3) Many APIs take in array/ImmutableArray/ReadOnlySpan, none of which are easily convertible without allocating from a list

Reply:
> I don't think the ability to get the underlying array from a List is exposed on netfx

Do we really need this ability? I thought we want to get a span. There is a ```CollectionsMarshal.AsSpan<T>(List<T>)``` for that.

> List doesn't directly support pooling

I think there are several examples how to make standard collections "poolable".

> Many APIs take in array/ImmutableArray/ReadOnlySpan, none of which are easily convertible without allocating from a list

Are we talking about collections in general, or in context of the problem that we are trying to address in this PR? It looks like we are looking for a list from which we can get a span. Well, there is already a list like that. Why would we want to invent something new?

Reply:
> There is a CollectionsMarshal.AsSpan<T>(List<T>) for that

Is that available on netfx? I didn't think it was.

> I think there are several examples how to make standard collections "poolable".

Yes, it's easy enough to make things poolable, it's just that it can be a bit of a hassle for users.

> Are we talking about collections in general, or in context of the problem that we are trying to address in this PR?

I'm fine just addressing the allocations in this PR, but I thought this mechanism would allow others to handle this situation nicely too.

Reply:
I wuold prefer to avoid 'List'.  One value in having custom types is that the 'standard' types stick out like a sore thumb as 'likely wrong'.  When we don't use List anywhere else (and we've actually been removing straggling usages of it in places that keep showing up in investigations), it's nice to not start reintroducing it.

Reply:
> I do not share this this view of the world.

Sure.  But we need an approach that works for both worlds.  An exception for IDE makes life harder there.  Whereas compiler would need no exception here, as it already uses custom types.  I get that you may be comfortable with this.  But i do ask that you consider the areas that partner teams have found valuable in wrangling complexity and dealing with this space :)

> So far I've seen no proof that the new type is "better" than List<T> in the specific scenario. 

I gave a reason for this above.  The new type is better in that it clearly indicates intent, and indicates the purpose of use.  It does not then carve out permission to use a type we do not want reused in other places, allowing us to still stick with our general desire to not be using List.  This may not be a better rule for the needs of the compiler team.  However, this area is a bit strange in that it appears generally owned by the IDE team, while layered into the compiler codebase.  I'm not sure how best to handle such a case.  With ownership by IDE, should we follow IDE prescriptions on coding patterns?  Or since layered into compiler, should we follow what the compiler would do in the pieces they more traditionally own?

We don't have a lot of precedence here.  But my preference is to defer to whichever team expects to work in the area the most, and own the maintenance/future of it.  If that's more compiler, i would defer to its preferences.  If it's more IDE, then the same :)

Reply:
I wouldn't object if the type is added and used strictly in IDE layer, not in shared and not Compilers layer.

Reply:
If all we want is a "special" type name. There is an option of deriving from ```List<T>``` and using the derived type instead of the ```List<T>``` directly. Basically, I do not think we should be creating yet another list implementation when we already have a standard collection type that fits our needs.

Reply:
I mean fits in terms of behavior.

Reply:
I'd like to better understand what the proposed change Aleksey is suggesting. Is it:

1) Change the four AnalyzerExecutor.Execute*Actions to take in a List rather than the current IEnumerable (or ROS that the PR has currently)
2) Change AnalyzerDriver.DeclarationAnalysisData to declare an objectpool<list> and to use that to get DescendantNodesToAnalyze
3) Change invocations of AnalyzerExecutor.Execute*Actions to pass in a List. This will require some callers to copy into a temporary list as they have an ImmutableArray (see AnalyzerDriver.ExecuteDeclaringReferenceActions.executeOperationsActionsByKind)

I can try changing to this if there is a strong aversion to adding a new collection type (which appears to be the case). I don't love that those four methods would now take in a mutable type, or that we would need to do the copy from the ImmutableArray into a list, but it's all doable. Again, I liked the approach currently in this PR as I see a need for us to be able to get an array or ROS from a List or ArrayBuilder like object and we currently can't do that (with the exception that net core can do that for List)

Reply:
> I'd like to better understand what the proposed change Aleksey is suggesting. Is it:

I we really want to pass a span around and avoid copying data, I suggest using ```List<T>``` instead of introducing a custom collection implementation.

There are also alternatives to consider, like either passing ```ImmutableArray``s, or or passing ```ArrayBuilder```s, or creating a copy to get a span.

I would be fine with either approach as long as it addresses the problem we are trying to address, doesn't introduce an inacceptable perf downside elsewhere, and doesn't add a custom ```List<T>``` implementation.

Reply:
Still testing, but I went ahead and pushed out changes that remove the added collection type. As mentioned, there is some extra data being copied around, but it should avoid the allocations this PR concerns itself with.

I still do like the type that was introduced earlier in this PR and do see a value that it provides, but I'm sensitive that the compiler team might not want to add another data type used in their area.